### DISCUSSION TITLE
How to deploy contract with linked libraries or get contract factory in hardhat-viem

### DISCUSSION THREAD

**Discussion by Aliceonly** [NomicFoundation/hardhat#4626] [Troubleshooting] [answered]
Deploying contract with library Model , its usage in contract like:
`using Model for xxxxxx;`

In ethers , we call `ethers.getContractFactory({libraries: libraries})` to attach libraries

If we don't pass the libraries , viem panic in deployment :
`TransactionExecutionError: Invalid parameters were provided to the RPC method. Double check you have provided the correct parameters.`

How to pass libraries when deploying?

**Comment by Chlebamaticon** [comment, marked as answer]
Had the same problem. I've submitted [the following PR](https://github.com/NomicFoundation/hardhat/pull/4695) that will provide such support. However, for now I can only suggest copy pasting the code as a workaround and parse the bytecode on your own.

[You will need this function.](https://github.com/NomicFoundation/hardhat/pull/4695/files#diff-cec65c4fe6f7f51906befedb00cedfe445c653eb4bf5b32886f9671d6d9e1e07R99)
```typescript
export async function resolveBytecodeWithLinkedLibraries(
  artifact: Artifact,
  libraries: Libraries<viemT.Address>
): Promise<string>;
```
Once you have it you will only need to parse bytecode and deploy it using `deployContract`.
```ts
const artifact = await hre.artifacts.readArtifact(
  contractName,
);
const publicClient = await hre.viem.getPublicClient();
const [client] = await hre.viem.getWalletClients();
const hash = await client.deployContract({
  abi: artifact.abi,
  bytecode: resolveBytecodeWithLinkedLibraries(artifact, { library: '0x00...00' }),
  args: [],
});
return publicClient.waitForTransactionReceipt({ hash });
```
